### DISCUSSION TITLE
Symbols missing from cgnsdll.dll

### DISCUSSION THREAD

**Discussion by gmic21** [CGNS/CGNS#768] [CGNStalk]
I have built a CGNS DLL using v4.3.0 and cmake. When I go to link to my program I get missing symbols that existed in prior CGNS version DLL, which I had built with configure and nmake. I have verified the absence of the symbold in the v4.3.0 DLL build and their presence in the prior version DLL build with dependency walker so this is not a calling name issue. The symbols in question are all FORTRAN callable CG_*_F symbols such as CG_NBASES_F, CG_NCOORDS_F, CG_NSECTIONS_F etc.

When I build the CGNS DLL I get the following output in MSVS 2022 project generated by cmake

Build started at 11:49 AM...
1>------ Build started: Project: ZERO_CHECK, Configuration: Release x64 ------
1>1>Checking Build System
2>------ Build started: Project: cgns_f2c (IFORT), Configuration: Release x64 ------
Compiling with Intel® Fortran Compiler Classic 2021.9.0 [Intel(R) 64]...
cgns_f.F90
Creating library...

Build log written to  "file://{deleted}/cgns/cgnslib_4.3.0/Release/src/cgns_f2c.dir/Release/BuildLog.htm"
cgns_f2c - 0 error(s), 0 warning(s)
3>------ Build started: Project: cgns_shared, Configuration: Release x64 ------
3>Building Custom Rule {deleted}/cgns/cgnslib_4.3.0/src/CMakeLists.txt
3>cl : command line  warning D9002: ignoring unknown option '/fpp'
3>cl : command line  warning D9002: ignoring unknown option '/libs:dll'
3>cl : command line  warning D9002: ignoring unknown option '/threads'
3>cgns_error.c
3>cgns_internals.c
3>cgns_io.c
3>cgnslib.c
3>cg_hashmap.c
3>ADF_interface.c
3>ADF_internals.c
3>ADFH.c
3>Generating Code...
3>{deleted}\cgns\cgnslib_4.3.0\src\adfh\ADFH.c(2253,1): warning C4700: uninitialized local variable 'bkg' used
3>{deleted}\cgns\cgnslib_4.3.0\src\adfh\ADFH.c(2253,1): warning C4700: uninitialized local variable 'tconv' used
3>{deleted}\cgns\cgnslib_4.3.0\src\adfh\ADFH.c(2342,1): warning C4700: uninitialized local variable 'bkg' used
3>{deleted}\cgns\cgnslib_4.3.0\src\adfh\ADFH.c(2342,1): warning C4700: uninitialized local variable 'tconv' used
3>cl : command line  warning D9002: ignoring unknown option '/fpp'
3>cl : command line  warning D9002: ignoring unknown option '/libs:dll'
3>cl : command line  warning D9002: ignoring unknown option '/threads'
3>cg_ftoc.c
3>cgio_ftoc.c
3>Generating Code...
3>   Creating library {deleted}/cgns/cgnslib_4.3.0/Release/src/Release/cgnsdll.lib and object {deleted}/cgns/cgnslib_4.3.0/Release/src/Release/cgnsdll.exp
3>cgns_shared.vcxproj -> {deleted}\cgns\cgnslib_4.3.0\Release\src\Release\cgnsdll.dll
3>Done building project "cgns_shared.vcxproj".
========== Build: 3 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 11:49 AM and took 05.519 seconds ==========

CG_NBASES_F is in cg_ftoc.c and the binding to the FORTRAN call is in cgns_f.F90, both of which get built and, presumably, linked. Here is an image of the cgns_f.obj file being included in the object libraries for the cgns_shared project and the source cg_ftoc.c is there as well.

![image](https://github.com/CGNS/CGNS/assets/171726996/aec86346-5865-43bb-83e5-e299326deab9)

Yet, CG_NBASES_F is not there

![image](https://github.com/CGNS/CGNS/assets/171726996/7f4bf129-9933-4dcf-b4bd-7b5c4030fb36)

How could these symbols not be in the resulting DLL?

**Comment by gmic21** [comment]
I figured out that naming is indeed the issue. While CG_NBASES_F does not exist, cg_nbases_f does exist. My main application requires UPPERCASE naming because so many other linked libraries use it. When I run cmake I see that it detects it should use an UPPERCASE naming convention to Fortran.

-- Setting Fortran enable option
-- Fortran enable option set to: ON
-- Fortran option enabled
-- The Fortran compiler identification is Intel 2021.9.0.20230302
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Determine Intel Fortran Compiler Implicit Link Path
-- Determine Intel Fortran Compiler Implicit Link Path - done
-- Check for working Fortran compiler: C:/Program Files (x86)/Intel/oneAPI/compiler/2023.1.0/windows/bin/intel64/ifort.exe - skipped
-- Detecting Fortran/C Interface
-- Detecting Fortran/C Interface - Found GLOBAL and MODULE mangling
-- Fortran name mangling convention: UPPERCASE

Why are the FORTRAN calls not all in UPPERCASE?

**Reply by MicK7** [reply]
The name mangling is not done in cg_ftoc for this function. The function is defined in cgns_f.F90 directly using iso C mapping.
The Fortran API is a mess has part is old fortran and part is modern fortran 90 or 2003.
If you have lowercase mangling you do not see the difference but in other case it is broken.

**Reply by gmic21** [reply]
Thank you for your confirmation that I have to do something to get this to work. I did edit the routines I need to call from FORTRAN and I think I am moving again because I get a build of the DLL. Not too big of a deal to do but figuring out _what_ I had to do was the hard part. The fact that some were completed already was helpful to show me what I had to do, but also confusing because I could not understand if the mix of syntax was intentional for a reason I did not understand. But what I did is inconsistent with your comments. In cg_ftoc.c I made changes like this

what was this:

CGNSDLL void cg_nzones_f(cgint_f *fn, cgint_f *B, cgint_f *nzones, cgint_f *ier)

becomes this:

CGNSDLL void FMNAME(cg_nzones_f, CG_NZONES_F) (cgint_f *fn, cgint_f *B, cgint_f *nzones, cgint_f *ier)

In cgns_f.F90 I commented out the BIND so that this:

     SUBROUTINE cg_nbases_f(fn, nbases, ier) BIND(C, NAME="cg_nbases_f")

becomes this:

     SUBROUTINE cg_nbases_f(fn, nbases, ier) !BIND(C, NAME="cg_nbases_f")

I did this for any routine that I need to call from FORTRAN.

I have other DLL issues not related to this so I am still not quite at a full test through my application to know whether this is working.

I am then going to call with an UPPERCASE calling convention. Does the above make sense?

**Reply by brtnfld** [reply]
The functions that don't pass a string use BIND(C), as the standard does not allow Fortran strings to be passed to C. This has since been addressed in F2023, so we should be able to detect if this is the case and use BIND(C) for those other cases. We are trying to eliminate the uppercase and lowercase issues by using bind(c), so your fix goes backward.

 I'll need to look into why the DLL definitions in cgns_fortrandll.def are not working, as, for example, cg_nbases_f is defined there. We might have missed adding some newer APIs to the file. We lack testing CGNS with Fortran on Windows; we are trying to address that issue.

**Reply by brtnfld** [reply]
Did all the CGNS tests pass for your installation?
Can you try modifying your CMake build? This configuration works for me using OneAPI.

```
cmake -G "NMake Makefiles" ^
 -D CMAKE_C_COMPILER=icx -D CMAKE_Fortran_COMPILER=ifx ^
  -D CMAKE_PREFIX_PATH:PATH="C:\Users\Public\Downloads\IntelBuild_serial\build\hdf5\cmake" ^
  -D CGNS_ENABLE_FORTRAN:BOOL=ON ^
  -D CMAKE_BUILD_TYPE:STRING="Release" ^
  -D CGNS_BUILD_SHARED:BOOL=ON ^
  -D CGNS_USE_SHARED:BOOL=ON ^
  -D CGNS_ENABLE_LEGACY:BOOL=ON ^
  -D CGNS_USE_SHARED:BOOL=ON ^
  -D CGNS_ENABLE_LFS:BOOL="ON" ^
  -D CGNS_ENABLE_64BIT=BOOL=ON ^
  -D CMAKE_VERBOSE_MAKEFILE:BOOL=ON ^
  -D CMAKE_C_FLAGS:STRING="-DH5_BUILT_AS_DYNAMIC_LIB -DUPPERCASE /MD /Qstd=c99" ^
  -D CMAKE_Fortran_FLAGS:STRING="-DH5_BUILT_AS_DYNAMIC_LIB -DUPPERCASE /MD -fpp" ^
  -D CGNS_ENABLE_TESTS:BOOL=ON ^
  -D CGNS_ENABLE_HDF5:BOOL=ON ^
  -D CGNS_BUILD_CGNSTOOLS:BOOL=OFF ^
  -D HDF5_NEED_ZLIB:BOOL=OFF ^
  -D HDF5_NEED_SZIP:BOOL=OFF ^
  -D CGNS_ENABLE_PARALLEL:BOOL=ON^
  -D MPIEXEC_PREFLAGS:STRING="-localonly"^
  -D MPIEXEC:STRING="mpiexec"^
  -D HDF5_NEED_MPI:BOOL=ON^
   ..
cmake --build .

```

**Comment by gmic21** [comment]
Scot,

Some of the tests were quarantined by my antivirus so I’m getting that straightened out. If I open that tests folder all these warning bells go off and my IT calls me in alarm. I have to submit them to our antivirus for analysis before I can run them

I can build a cgnsdll.dll with your configuration below and no code changes, just like I could with the MSVS files. The issue is that linking to the generated library generates symbols not found errors just the same. If I edit the calls I am using so that they are defined without BIND(C) just like the calls with strings, then I can link. I am unclear if there is some other way to fix this name mangling issue and still use the BIND(C).

But runs with my application (linked to the edited form that links) fail. I have not added any kind of error recovery yet to figure out what is happening.  It fails during read of a file that runs fine with the same application code but 3.2.1 of CGNS. I have not familiarized myself much with the error reporting so there is probably some error message waiting for me to query. Some calls do check for errors but not all.

For now I am going to have to punt and backpedal to 3.2.1 and then come back to this. It is not entirely clear to me what needs to be done to the FORTRAN interface to get it working with the BIND(C). If what needs to happen is a lot of mundane legwork, then I am happy to do that if it is made clear to me what syntax to use and where. Maybe an example of what needs to be changed to what on a couple of interfaces. If it’s clear I am happy to knock that out, maybe with some scripting. I really would like this to be effortless when I pull the latest version of CGNS, so I’m happy to help where I can.

Best
Greg

Gregory J Michels
Sigmadyne, Inc.
1200C Scottsville Rd., Suite 350
Rochester, NY 14624, USA
P: (585)235-6892
E: ***@***.******@***.***>

***@***.***

**Comment by brtnfld** [comment]
Created Issue #773 for further discussion.
